dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers/map-fee-record-payment-entity-groups-to-fee-record-payment-groups.ts

```diff
@@ -1,56 +1,123 @@
+import { FeeRecordPaymentGroup, FeeRecordReconciledByUser } from '../../../../../types/utilisation-reports';
+import { TfmUsersRepo } from '../../../../../repositories/tfm-users-repo';
 
+/**
+ * Gets the status for a fee record payment entity group
+ * @param feeRecordEntitiesInGroup - The fee record entities in a group
+ * @returns The status of the group
+ */
 const getStatusForGroupOfFeeRecords = (feeRecordEntitiesInGroup: FeeRecordEntity[]): FeeRecordStatus => {
   if (feeRecordEntitiesInGroup.some((feeRecordEntity) => feeRecordEntity.status === 'READY_TO_KEY')) {
     return 'READY_TO_KEY';
   }
-
   return feeRecordEntitiesInGroup[0].status;
 };
 
+/**
+ * Gets the most recently reconciled fee record from a list of fee records
+ * @param feeRecordsAtReconciledStatus - The fee records (with status RECONCILED)
+ * @returns The most recently reconciled fee record
+ */
+const getMostRecentlyReconciledFeeRecord = (feeRecordsAtReconciledStatus: FeeRecordEntity[]): FeeRecordEntity => {
+  if (feeRecordsAtReconciledStatus.some((feeRecord) => feeRecord.dateReconciled === null)) {
+    throw new Error('Expected list of fee records at RECONCILED status to all have a non-null date reconciled');
+  }
+  const feeRecordsSortedByDateReconciledDescending = orderBy(feeRecordsAtReconciledStatus, [(feeRecord) => feeRecord.dateReconciled!.getTime()], ['desc']);
+  return feeRecordsSortedByDateReconciledDescending[0];
+};
+
+/**
+ * Gets the reconciled by user for a fee record
+ * @param feeRecord - The fee record
+ * @returns The reconciled by user
+ */
+const getReconciledByUserForFeeRecord = async (feeRecord: FeeRecordEntity): Promise<FeeRecordReconciledByUser | null> => {
+  if (feeRecord.reconciledByUserId === null) {
+    return null;
+  }
+  const user = await TfmUsersRepo.findOneUserById(feeRecord.reconciledByUserId);
+  if (!user) {
+    console.info("Find to find the TFM user with id '%s' who reconciled a fee record payment group", feeRecord.reconciledByUserId);
+    return null;
+  }
+  return {
+    firstName: user.firstName,
+    lastName: user.lastName,
+  };
+};
+
+/**
+ * Gets the reconciliation data for the fee records in the group
+ * @param feeRecords - The fee records in the group
+ * @param groupStatus - The status of the fee record payment group
+ * @returns The reconciliation data
+ */
+const getReconciliationDataForFeeRecords = async (
+  feeRecords: FeeRecordEntity[],
+  groupStatus: FeeRecordStatus,
+): Promise<{ reconciledByUser: FeeRecordReconciledByUser | null; dateReconciled: Date | null }> => {
+  if (groupStatus !== FEE_RECORD_STATUS.RECONCILED) {
+    return { reconciledByUser: null, dateReconciled: null };
+  }
+
+  const mostRecentlyReconciledFeeRecord = getMostRecentlyReconciledFeeRecord(feeRecords);
+  const reconciledByUser = await getReconciledByUserForFeeRecord(mostRecentlyReconciledFeeRecord);
+  return {
+    reconciledByUser,
+    dateReconciled: mostRecentlyReconciledFeeRecord.dateReconciled,
+  };
+};
+
```

Comment:
Same as above, can't see any unit tests for the functions above here?

Reply:
That's because these functions all get tested as part of the testing of the exported function. There is no reason to export every function from this file just for the sake of more isolated testing as we really want to test our edge cases in the context of the main mapping function; not as smaller tests for these smaller functions

Reply:
Unfortunately I don't agree - see comment above RE testing.

Sorry but I don't see any reason for _not_ having tests for a function.

Other functions that consume such functions, shouldn't care about a child function's logic since they only call it.

With the current approach, if one of the "child" functions is changed and fails, it will be reported as an issue in the currently-only exported function. It will be tricky for someone to debug. Isolated tests prevent such issues.

Also, the functions could be used in other places in which case having tests in place already would be great and would save future effort.

Reply:
These functions in particular should _not_ ever be used by another function, hence the purposeful choice to not include them as an `export`. If a function is ever shared by another piece of functionality, we move it to the `helpers` directory and add a set of unit tests there (see the fee record payment group payment logic for example).

However, in this context, adding tests for every one of these functions actually causes more issues than it alleviates. The main purpose of these functions is to provide self-documenting code, so that a user reading the mapping function has can read each line as an action rather than having to infer the action from the implementation (ie. imperative). This means that such functions are purposefully designed to be able to be swapped out for a function with a different name and/or implementation as the situation calls for it. With the current approach, such a change would be caught in the unit test of the mapping function and could be updated there, possibly causing a few lines to change. If we test every function in this way, then renaming a function for the sake of documentation now causes you to have to update every test linked to that specific function, which is a much larger refactor and results in a less-developer friendly experience

Reply:
I'm very much of the opinion we shouldn't export functions solely for testing purposes, as it exposes code that is not intended to be exposed. Very happy for us to chat about this further offline @ttbarnes!

Reply:
>It is a recommended practice you write a unit test for any unit of functionality you create

I think this is an oversimplification of the principle behind unit testing - this is definitely a discussion that needs to be brought offline rather than continuing it in this thread

Reply:
Completely agree with @abhi-markan - every piece of logic/functionality should have a unit test.

The current approach I suspect does not cover the possible conditions/scenarios, test coverage is also probably low. We should be striving for 100% test coverage.

I understand the reasoning behind "don't export something just to test it, it would otherwise not be exposed", however by not doing this, it exposes more risk (by not having individual tests and makes the current test huge and hard to read/maintain.

In the future, someone could easily add an export and start using such functions if required.

Reply:
>test coverage is also probably low

Test coverage is very much not low using this approach. Again, this discussion should be taken offline

Reply:
@JoshBinns2000 What does the test coverage report currently say for this file?

We need full test coverage for individual functions.

Reply:
I think this thread really highlights that the scope of this function and file has grown quite a lot in this PR and the clarity and thus usefulness of the tests has diminished. I would propose we do the following:
- Pull the determining of the most recently reconciled fee record and the fetching of the user details out of this file. There is quite a lot of logic involved and I agree it would be worth testing this more directly (and perhaps this is highlighting some misplaced responsibility). How I imagine this looking @Josh is:
	- In `getUtilisationReportReconciliationDetails` we add a new line:
```
const feeRecordPaymentEntityGroups = getFeeRecordPaymentEntityGroupsFromFeeRecordEntities(feeRecords);

const groupsWithReconciledByUser = getReconciledByUsersForGroups(feeRecordPaymentEntities) // this is new

const feeRecordPaymentGroups = mapFeeRecordPaymentEntityGroupsToFeeRecordPaymentGroups(
facilityIdFilter ? filterFeeRecordPaymentEntityGroupsByFacilityId(groupsWithReconciledByUser, facilityIdFilter) : groupsWithReconciledByUser,
  );
```

(Where `getReconciledByUsersForGroups` returns something along the shape of `{ feeRecords: FeeRecordEntity[], payments: PaymentEntity[], reconciledByUser: TfmUser | null }`.)
- Pull out `getStatusForGroupOfFeeRecords` into a commonised helper with comprehensive dedicated unit tests as the logic here would be the same anywhere else we displayed a group of records with a status if we ever were to.
- Update the test for the mapper to mock these out so we don't need to test all the edge cases in the mapper tests since they will be covered by their own unit tests, which should simplify and make the mapper tests easier to read.

Give me a shout if you want to chat about any of this/any of that doesn't make sense!

Reply:
I have taken the above suggestion as inspiration and instead added helpers which operate on `FeeRecordPaymentEntityGroup` objects in the `fee-record-payment-entity-group.helper.ts` file